flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java

```diff
@@ -0,0 +1,229 @@
+            options.put("disable-lookup", "true");
+            options.put("runtime-source", "NewSource");
+            sourceTestStep.apply(tEnv, options);
+        }
+
+        final List<CompletableFuture<?>> futures = new ArrayList<>();
+        for (SinkTestStep sinkTestStep : program.getSetupSinkTestSteps()) {
+            final CompletableFuture<Object> future = new CompletableFuture<>();
+            futures.add(future);
+            final String tableName = sinkTestStep.name;
+            TestValuesTableFactory.registerLocalRawResultsObserver(
+                    tableName,
+                    (integer, strings) -> {
+                        final boolean canTakeSavepoint =
+                                sinkTestStep.expectedBeforeRestore.test(
+                                        TestValuesTableFactory.getRawResults(tableName));
+                        if (canTakeSavepoint) {
+                            future.complete(null);
+                        }
+                    });
+            final Map<String, String> options = new HashMap<>();
+            options.put("connector", "values");
+            options.put("disable-lookup", "true");
+            options.put("sink-insert-only", "false");
+            sinkTestStep.apply(tEnv, options);
+        }
+
+        final SqlTestStep sqlTestStep = program.getRunSqlTestStep();
+
+        final CompiledPlan compiledPlan = tEnv.compilePlanSql(sqlTestStep.sql);
+        compiledPlan.writeToFile(getPlanPath(program, getCurrentVersion()));
+
+        final TableResult tableResult = compiledPlan.execute();
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).get();
+        final JobClient jobClient = tableResult.getJobClient().get();
+        final String savepoint =
+                jobClient
+                        .stopWithSavepoint(false, tmpDir.toString(), SavepointFormatType.DEFAULT)
+                        .get();
+        CommonTestUtils.waitForJobStatus(jobClient, Collections.singletonList(JobStatus.FINISHED));
+        final Path savepointPath = Paths.get(new URI(savepoint));
+        final Path savepointDirPath = getSavepointPath(program, getCurrentVersion());
+        Files.createDirectories(savepointDirPath);
+        Files.move(savepointPath, savepointDirPath, StandardCopyOption.ATOMIC_MOVE);
+    }
+
+    @ParameterizedTest
+    @MethodSource("createSpecs")
+    void testRestore(int version, TableTestProgram program) throws Exception {
+        final EnvironmentSettings settings = EnvironmentSettings.inStreamingMode();
+        final SavepointRestoreSettings restoreSettings =
+                SavepointRestoreSettings.forPath(
+                        getSavepointPath(program, version).toString(), false, RestoreMode.NO_CLAIM);
+        SavepointRestoreSettings.toConfiguration(restoreSettings, settings.getConfiguration());
+        final TableEnvironment tEnv = TableEnvironment.create(settings);
+        tEnv.getConfig()
```

Comment:
maybe we should add a note that this test base is not able to test catalog objects?

Reply:
maybe it's better to use `PLAN_COMPILE_CATALOG_OBJECTS` to make the checked in JSON plans also smaller